New: Add forward scrubbing prevention (fix #323)

joe-replin opened the pull request:
[//]: # (Please title your PR according to eslint commit conventions)
[//]: # (See https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-changelog-eslint#eslint-convention for details)

[//]: # (Link the PR to the original issue)

[//]: # (Delete Fix, Update, New and/or Breaking sections as appropriate)
https://github.com/adaptlearning/adapt-contrib-media/pull/320 - was not a comprehensive solution for preventing forward scrubbing. Making another attempt.

### Fix
- Fixes #323 
- Fixes #333
- No pointer events prevents mouse and touch events.
- Prevent default blocks the scrub bar from being changed programmatically.
- aria-disabled makes the scrub bar non-focusable.
- Prevents events from firing when attempting to scrub passed a locked segment of the video.

### Update
- Adjusted `_preventForwardScrubbing` configuration option that prevents learners from skipping ahead in audio/video content until they have viewed/listened to it completely at least once
- Learners can still skip backwards to any previously viewed portion of the media
- Visual blocker overlay appears on the progress bar to indicate the restricted area. Animation provides feedback when learners attempt to scrub forward beyond their max viewed position
- Keyboard navigation (ArrowRight, End, PageDown) is also restricted when at the maximum viewed position
- Restriction is automatically lifted once the component is marked as complete
- Event listener management for forward scrubbing prevention to ensure proper cleanup on component removal
- README, Schemas & Example

### New

[//]: # (List appropriate steps for testing if needed)
### Testing
1. Enable `_preventForwardScrubbing: true` in a media component
2. Set `_setCompletionOn: "ended"`
3. Play the media and verify:
  - You cannot click or drag the progress bar ahead of the current max viewed position
  - You can click/drag backwards to any previously viewed position
  - Arrow keys (ArrowRight, End, PageDown) do not skip ahead when at max viewed position
  - A visual blocker appears on the progress bar showing the restricted area
  - Flash animation displays when attempting to scrub forward
  - After reaching the end and completing the component, all restrictions are removed
4. Test with both audio and video media
5. Verify proper cleanup by navigating away from the page and back

[//]: # (Mention any other dependencies)

joe-replin committed:
Update mediaView.js

joe-replin committed:
Update mediaelementplayer.less

joe-replin committed:
Update mediaView.js

joe-replin committed:
Update mediaView.js

swashbuck commented on js/mediaView.js:
```diff
@@ -536,10 +536,125 @@ class MediaView extends ComponentView {
+
+    // Prevent forward seeking and navigate to maxViewed
+    player.addEventListener('seeking', () => {
+      if (player.currentTime > getMaxViewed() + 0.25) {
```
Could return early here

swashbuck commented on js/mediaView.js:
```diff
@@ -536,10 +536,125 @@ class MediaView extends ComponentView {
+  setupScrubBlockerEvents(player, scrubBlocker, getMaxViewed, setMaxViewed, getSuppress, setSuppress) {
+  // Update progress and blocker size
+    player.addEventListener('timeupdate', () => {
+      if (!getSuppress()) {
```
Return early

swashbuck commented on js/mediaView.js:
```diff
@@ -536,10 +536,125 @@ class MediaView extends ComponentView {
+      const maxViewed = this.model.get('_maxViewed') || 0;
+
+      // If clicking ahead of maxViewed, navigate to maxViewed and flash
+      if (clickTime > maxViewed + 0.25) {
```
Return early

swashbuck commented on js/mediaView.js:
```diff
@@ -536,10 +536,125 @@ class MediaView extends ComponentView {
+    // Prevent keyboard forward navigation
+    player.addEventListener('keydown', (e) => {
+      const forwardKeys = ['ArrowRight', 'End', 'PageDown'];
+      if (forwardKeys.includes(e.code) && player.currentTime >= getMaxViewed()) {
```
Return early

swashbuck commented on js/mediaView.js:
```diff
@@ -536,10 +536,125 @@ class MediaView extends ComponentView {
+    scrubBlocker.className = 'mejs__time-slider-blocker';
+
+    const flashBlockedOverlay = () => {
+      scrubBlocker.classList.add('mejs__time-slider-blocker-error');
+      setTimeout(() => {
+        scrubBlocker.classList.remove('mejs__time-slider-blocker-error');
+      }, 150);
+    };
```
This code appears three times in this file. Suggest making it a function and calling that each time.

joe-replin replied:
> This identical code appears three times in this file. Suggest making it a function and calling that each time.
Done. Good spot.

> Also, I'm wondering about the accessibility of the 150ms red flash. If someone clicks or key presses multiple times, the red can flash in quick succession. Maybe this should only flash a maximum of once every 1 second?
> 
> https://www.w3.org/WAI/WCAG22/Understanding/three-flashes-or-below-threshold.html#dfn-general-flash-and-red-flash-thresholds
I'm not so sure about this call. It's not a continuous animation and the user would need to rapid-fire click to create a violation of the rule, which is highly unlikely. Also, the goal of the rule is to avoid triggering seizures. The size of the square doesn't seem large enough to fall under WCAG 2.3.1? Happy to implement something but I'm no so sure it is needed. I could be swayed the other direction pretty easily though.

joe-replin committed:
Update mediaView.js

joe-replin committed:
Update mediaView.js

joe-replin commented:
Noting here that this PR has been extensively tested and is currently a part of a large production release with no issues.

swashbuck commented on less/mediaelementplayer.less:
```diff
@@ -511,6 +511,23 @@ Reference: http://blog.rrwd.nl/2015/04/04/the-screen-reader-text-class-why-and-h
+    background-color: fade(@validation-error, 30%);
+  }
+}
+
```
Will you move these styles to _mep-overrides.less_? The _mediaelementplayer.less_ file is a direct copy of the library's [Less](https://github.com/mediaelement/mediaelement/blob/7.0.7/src/css/mediaelementplayer.css) file and should not be modified. Otherwise, the styles may be overridden during a library update.

joe-replin replied:
Done!

joe-replin committed:
Adjust LESS location

cahirodoherty-learningpool reviewed (approved):
👀

oliverfoster commented on js/mediaView.js:
```diff
@@ -536,10 +536,115 @@ class MediaView extends ComponentView {
+    if (!$slider.length) return;
+
+    let maxViewed = this.model.get('_maxViewed') ?? 0;
+    let suppress = false;
```
These two variables and their associated functions seem completely mad, lots of unnecessarily complex code.

There are two callback functions each, passed into `setupScrubBlockerEvents`, `getMaxViewed`, `setMaxViewed`, `getSupress` and `setSupress`, both of which just toggle these local boolean variables.

Why not remove all four callback functions, the four arguments in `setupScrubBlockerEvents` and just set the boolean value on the MediaView instance as a `this.maxViewed` and `this.suppress` property instead?

joe-replin replied:
Appreciate you calling me out on it. It was a convoluted approach. I will revise.

oliverfoster commented on js/mediaView.js:
```diff
@@ -536,10 +536,115 @@ class MediaView extends ComponentView {
+
+    // Prevent keyboard forward navigation
+    player.addEventListener('keydown', (e) => {
+      const forwardKeys = ['ArrowRight', 'End', 'PageDown'];
```
This should be at the top of the file as `const FORWARD_KEYS = ['ArrowRight', 'End', 'PageDown'];` 

Currently, a keydown event occurs on the mediaelement or its descendants and then this array and its strings are created every time with identical values.

oliverfoster commented on js/mediaView.js:
```diff
@@ -536,10 +536,115 @@ class MediaView extends ComponentView {
-      seeking: this.onMediaElementSeeking,
-      timeupdate: this.onMediaElementTimeUpdate
+
+    const player = this.mediaElement;
```
The player variable does not need to be passed through to each function. As the functions are on the view, they can refer to `this.mediaElement`. This code is written as if the functions are not view instance functions.

joe-replin committed:
Clarify _setCompletionOn comment in example.json

Updated comment to clarify usage of _setCompletionOn.

joe-replin committed:
Forward scrubbing prevention

joe-replin committed:
Media scrubbing prevention settings

Updated title and help text for preventing media scrubbing.

joe-replin committed:
_preventForwardScrubbing property details

Updated title and description for _preventForwardScrubbing property to clarify its functionality.

joe-replin committed:
_preventForwardScrubbing behavior in README

Updated the description of the _preventForwardScrubbing setting to clarify its behavior regarding skipping ahead in media.

joe-replin committed:
Update README with native player controls note

Added note about native player controls for scrubbing restriction.

cahirodoherty-learningpool reviewed (approved):
👀

swashbuck reviewed (approved):
👍

oliverfoster commented on js/mediaView.js:
```diff
@@ -445,11 +447,39 @@ class MediaView extends ComponentView {
+      }
+      if (this._onScrubEnded) {
+        this.mediaElement.removeEventListener('ended', this._onScrubEnded);
+      }
```
These checks are all a consequence of assigning the functions to the instance at runtime rather that assigning the functions to the class at compile time.

```suggestion
      this.mediaElement.removeEventListener('timeupdate', this._onScrubTimeUpdate);
      this.mediaElement.removeEventListener('seeking', this._onScrubSeeking);
      this.mediaElement.removeEventListener('keydown', this._onScrubKeyDown);
      this.mediaElement.removeEventListener('ended', this._onScrubEnded);
```

oliverfoster commented on js/mediaView.js:
```diff
@@ -535,11 +565,124 @@ class MediaView extends ComponentView {
+      this._scrubBlocker?.remove();
+    };
+
+    this.mediaElement.addEventListener('timeupdate', this._onScrubTimeUpdate);
```
All of these nested functions should be assigned to the media prototype at compile time rather than run-time.

```suggestion
    this._scrubBlocker = this.createScrubBlocker($slider[0]);

    this.setupScrubBlockerEvents();

    // Initialize the blocker size
    this.updateScrubBlocker();
  }
  
  _onBlockerPointerDown (e) {
    e.preventDefault();
    e.stopImmediatePropagation();
    this.flashBlockedOverlay(scrubBlocker);
  }

  _onSliderClick (e) {
    const rect = sliderElement.getBoundingClientRect();
    const clickX = e.clientX - rect.left;
    const sliderWidth = rect.width;
    const clickPercent = clickX / sliderWidth;
    const duration = this.mediaElement.duration;
    const clickTime = clickPercent * duration;
    const isClickingAhead = clickTime > this._maxViewed + 0.25;

    if (!isClickingAhead) return;

    e.preventDefault();
    e.stopImmediatePropagation();
    this.mediaElement.currentTime = this._maxViewed;
    this.flashBlockedOverlay(scrubBlocker);
  }

  createScrubBlocker(sliderElement) {
    const scrubBlocker = document.createElement('span');
    scrubBlocker.className = 'mejs__time-slider-blocker';

    scrubBlocker.addEventListener('pointerdown', this._onBlockerPointerDown);
    sliderElement.addEventListener('click', this._onSliderClick);

    sliderElement.style.position = 'relative';
    sliderElement.appendChild(scrubBlocker);
    sliderElement.setAttribute('aria-disabled', 'true');

    return scrubBlocker;
  }
  
  onScrubTimeUpdate () {
    if (this._suppressSeek) return;
    this._maxViewed = Math.max(this._maxViewed, this.mediaElement.currentTime);
    this.model.set('_maxViewed', this._maxViewed);
    this.updateScrubBlocker();
  }

  // Prevent forward seeking and navigate to maxViewed
  onScrubSeeking () {
    const isSeekingAhead = this.mediaElement.currentTime > this._maxViewed + 0.25;
    if (!isSeekingAhead) return;

    this._suppressSeek = true;
    this.mediaElement.currentTime = this._maxViewed;
    this._suppressSeek = false;

    this.flashBlockedOverlay(this._scrubBlocker);
    this._showBlockedScrubMessage?.();
  }

  // Prevent keyboard forward navigation
  _onScrubKeyDown (e) {
    const isForwardKey = FORWARD_SCRUBBING_KEYS.includes(e.code);
    const isAtMaxViewed = this.mediaElement.currentTime >= this._maxViewed;
    const shouldPrevent = isForwardKey && isAtMaxViewed;

    if (!shouldPrevent) return;

    e.preventDefault();
    this.mediaElement.currentTime = this._maxViewed;
    this.flashBlockedOverlay(this._scrubBlocker);
  }

  _onScrubEnded () {
    this.model.set('_isComplete', true);
    this._scrubBlocker?.remove();
  }

  setupScrubBlockerEvents() {
    // Update progress and blocker size
    this.mediaElement.addEventListener('timeupdate', this._onScrubTimeUpdate);
```

oliverfoster commented on js/mediaView.js:
```diff
@@ -535,11 +565,124 @@ class MediaView extends ComponentView {
+      this._suppressSeek = false;
+
+      this.flashBlockedOverlay(this._scrubBlocker);
+      this._showBlockedScrubMessage?.();
```
This function doesn't exist.
```
      this._showBlockedScrubMessage?.();
```

joe-replin replied:
done

joe-replin committed:
Update js/mediaView.js

Co-authored-by: Oliver Foster <7974663+oliverfoster@users.noreply.github.com>

joe-replin committed:
Update js/mediaView.js

Co-authored-by: Oliver Foster <7974663+oliverfoster@users.noreply.github.com>

joe-replin committed:
Refactor media scrub blocker and captions handling
